conda_lock/models/lock_spec.py

```diff
@@ -24,23 +28,141 @@ class _BaseDependency(StrictModel):
     def sorted_extras(cls, v: List[str]) -> List[str]:
         return sorted(v)
 
+    def _merge_base(self, other: _BaseDependency) -> _BaseDependency:
+        if other is None:
+            return self
+        if (
+            self.name != other.name
+            or self.manager != other.manager
+            or self.category != other.category
+        ):
+            raise ValueError(
+                "Cannot merge incompatible dependencies: {self} != {other}"
+            )
+        return _BaseDependency(
+            name=self.name,
+            manager=self.manager,
+            category=self.category,
+            extras=list(set(self.extras + other.extras)),
+        )
+
 
 class VersionedDependency(_BaseDependency):
     version: str
     build: Optional[str] = None
     conda_channel: Optional[str] = None
 
+    @staticmethod
+    def _merge_matchspecs(
+        matchspec1: Optional[str],
+        matchspec2: Optional[str],
+        combine_constraints: bool = True,
+    ) -> Optional[str]:
+        if matchspec1 == matchspec2:
+            return matchspec1
+        if matchspec1 is None or matchspec1 == "":
+            return matchspec2
+        if matchspec2 is None or matchspec2 == "":
+            return matchspec1
+        if fnmatchcase(matchspec1, matchspec2):
+            return matchspec1
+        if fnmatchcase(matchspec2, matchspec1):
+            return matchspec2
+        if not combine_constraints:
+            raise ValueError(
+                f"Found incompatible constraint {matchspec1}, {matchspec2}"
+            )
+        return f"{matchspec1},{matchspec2}"
+
+    def merge(self, other: Optional[VersionedDependency]) -> VersionedDependency:
+        if other is None:
+            return self
+
+        if (
+            self.conda_channel is not None
+            and other.conda_channel is not None
+            and self.conda_channel != other.conda_channel
+        ):
+            raise ValueError(
+                f"VersionedDependency has two different conda_channels:\n{self}\n{other}"
+            )
+        merged_base = self._merge_base(other)
+        try:
+            build = self._merge_matchspecs(
+                self.build, other.build, combine_constraints=False
+            )
+        except ValueError as exc:
+            raise ValueError(
+                f"Unsupported usage of two incompatible builds for same dependency {self}, {other}"
+            ) from exc
+
+        return VersionedDependency(
+            name=merged_base.name,
+            manager=merged_base.manager,
+            category=merged_base.category,
+            extras=merged_base.extras,
+            version=self._merge_matchspecs(self.version, other.version),  # type: ignore
```

Comment:
We shouldn't need to use `# type: ignore` for this. We're missing a comprehensive treatment of corner cases for `None`. Namely, we know that the first argument in this line is not `None` and thus the return type is also not `None`. So some possibilities off the top of my head:
* Require the first argument of `_merge_matchspecs` to be not `None`, and same for the return type
* Raise an error in `_merge_matchspecs` if both arguments are `None`
* Allow `_merge_matchspecs` to return `None` but deal with this case.